### ISSUE TITLE
Time zone context manager?

### ISSUE THREAD

**Issue by pganssle** [spulec/freezegun#240]
In dateutil/dateutil#438, I've proposed adding a `testing` submodule to `dateutil` to allow some of the dateutil-specific test stuff I've created to be used in general. Another option was to spin these things off into a separate library, or to move them into `freezegun`.

The big one is a context manager which changes the effective system time zone, [`TZContextBase`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L95), [`TZEnvContext`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L154) and [`TZWinContext`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L178). Right now these actually change the time zone for the environment, but I'm thinking they can be reworked to use mocks instead (or have a flavor that mocks and a flavor that doesn't).

It would also be a simple matter to create decorator versions of these.

**Comment by spulec** [comment]
Yeah, that makes sense to me and seems like it would be a good fit.

Let me know if I can help.

**Comment by boxed** [comment]
I'm confused.. doesn't freezegun already allow you to do this? Freeze to the current time with tick=True and the timezone you want?

**Comment by pganssle** [comment]
@boxed I don't believe so.

This is actually made somewhat more difficult by the fact that updating the system time zone while a program is running does not always do what you would hope, particularly under Windows.

**Comment by boxed** [comment]
I forgot I wrote that comment... yea, freezegun is pretty broken with respects to timezones right now. There are many issues, but what I described above is probably what it _should_ do. It's just that localtime, gmtime, mktime and more aren't properly monkey patched. I've made an attempt to fix this but it's quite a big job. I would like to land some refactoring PRs first before tackling this.

**Comment by pganssle** [comment]
@boxed This issue is *mostly* unrelated to that. The patching of localtime, gmtime and mktime may be useful for the time-freezing stuff, but this issue is more about changing what the local equivalent of `TZ` is.

It mostly matters if you are using `dateutil.tz.tzlocal()`, `dateutil.tz.tzwinlocal()` or some other mechanism for retrieving the local time zone as a `tzinfo` instance, though obviously changing the effective local time is *also* likely useful.
